File: backend/src/test/java/com/ssafy/yorr/room/initializer/StaleRoomCleanerTest.java
package com.ssafy.yorr.room.initializer;

import com.ssafy.yorr.room.dto.RoomPhase;
import com.ssafy.yorr.room.dto.RoomSnapshot;
import com.ssafy.yorr.room.service.RoomCreateService;
import com.ssafy.yorr.room.service.RoomService;
import org.junit.jupiter.api.Test;

import java.util.List;
import java.util.Set;

import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

class StaleRoomCleanerTest {

    private final RoomCreateService roomCreateService = mock(RoomCreateService.class);
    private final RoomService roomService = mock(RoomService.class);
    private final StaleRoomCleaner cleaner = new StaleRoomCleaner(roomCreateService, roomService);

    /**
     * 이전 구현은 부팅 때 모든 방을 지워, 배포마다 살아있는 대기실까지 전멸시켰다.
     * 라운드 상태 없이도 정상 동작하는 방은 건드리지 않아야 한다.
     */
    @Test
    void keepsLobbyAndFinishedRoomsOnStartup() {
        when(roomCreateService.getAllRoomNumbers()).thenReturn(Set.of("lobby-room", "finished-room"));
        when(roomService.getSnapshot("lobby-room")).thenReturn(snapshot("lobby-room", RoomPhase.LOBBY));
        when(roomService.getSnapshot("finished-room"))
                .thenReturn(snapshot("finished-room", RoomPhase.FINISHED));

        cleaner.closeUnrecoverableGamesOnStartup();

        verify(roomService, never()).close("lobby-room");
        verify(roomService, never()).close("finished-room");
    }

    /**
     * 라운드 상태는 인메모리라 재시작으로 사라진다. Redis만 PLAYING으로 남은 방은 진행할 턴이
     * 없는데 JOIN이 game_started로 참가를 막아, TTL이 끝날 때까지 아무도 못 들어가는 방이 된다.
     */
    @Test
    void closesPlayingRoomsThatCanNoLongerContinue() {
        when(roomCreateService.getAllRoomNumbers()).thenReturn(Set.of("playing-room", "lobby-room"));
        when(roomService.getSnapshot("playing-room"))
                .thenReturn(snapshot("playing-room", RoomPhase.PLAYING));
        when(roomService.getSnapshot("lobby-room")).thenReturn(snapshot("lobby-room", RoomPhase.LOBBY));

        cleaner.closeUnrecoverableGamesOnStartup();

        verify(roomService).close("playing-room");
        verify(roomService, never()).close("lobby-room");
    }

    /** 이미 만료돼 phase를 못 읽는 방은 닫을 것도 없다 — close를 부르지 않는다. */
    @Test
    void ignoresRoomsThatAreAlreadyGone() {
        when(roomCreateService.getAllRoomNumbers()).thenReturn(Set.of("gone-room"));
        when(roomService.getSnapshot("gone-room")).thenReturn(RoomSnapshot.notFound("gone-room"));

        cleaner.closeUnrecoverableGamesOnStartup();

        verify(roomService, never()).close("gone-room");
    }

    private static RoomSnapshot snapshot(String roomCode, RoomPhase phase) {
        return new RoomSnapshot(roomCode, null, "host", phase, 6, List.of());
    }
}

File: backend/src/room/__tests__/staleRoomCleaner.test.ts
import { expect, it } from 'vitest'
import { describeRedis, useRedis } from '../../../test/redisHarness.js'
import type { UserIdentity } from '../../user/session.js'
import { RoomService } from '../roomService.js'
import { closeUnrecoverableGamesOnStartup } from '../staleRoomCleaner.js'

const HOST: UserIdentity = { userId: 'player-1', nickname: '호스트', type: 'GUEST' }

describeRedis('closeUnrecoverableGamesOnStartup', () => {
  const redis = useRedis()

  const openRoom = async (rooms: RoomService): Promise<string> => {
    const roomCode = await rooms.createRoom(6, HOST.userId, 'YACHT_DICE')
    await rooms.join(roomCode, HOST)
    return roomCode
  }

  /**
   * 이전 구현은 부팅 때 모든 방을 지워 배포마다 살아 있는 대기실까지 전멸시켰다.
   * 라운드 상태 없이도 정상 동작하는 방은 건드리지 않아야 한다.
   */
  it('대기실과 끝난 방은 그대로 둔다', async () => {
    const rooms = new RoomService(redis())
    const lobby = await openRoom(rooms)
    const finished = await openRoom(rooms)
    await rooms.startGame(finished, 1)
    await redis().hset(`room:${finished}`, 'phase', 'FINISHED')

    expect(await closeUnrecoverableGamesOnStartup(rooms)).toBe(0)

    expect((await rooms.getSnapshot(lobby)).phase).toBe('LOBBY')
    expect((await rooms.getSnapshot(finished)).phase).toBe('FINISHED')
  })

  /**
   * 마감 타이머는 인메모리라 재시작으로 사라진다. Redis만 PLAYING으로 남은 방은
   * 진행할 턴이 없는데 JOIN이 `game_started`로 참가까지 막는다.
   */
  it('이어갈 수 없는 진행 중 방을 닫는다', async () => {
    const rooms = new RoomService(redis())
    const playing = await openRoom(rooms)
    const lobby = await openRoom(rooms)
    await rooms.startGame(playing, 1)

    expect(await closeUnrecoverableGamesOnStartup(rooms)).toBe(1)

    expect((await rooms.getSnapshot(playing)).phase).toBeNull()
    expect((await rooms.getSnapshot(lobby)).phase).toBe('LOBBY')
  })

  it('이미 만료돼 phase를 못 읽는 방은 닫을 것도 없다', async () => {
    const rooms = new RoomService(redis())

    expect(await closeUnrecoverableGamesOnStartup(rooms)).toBe(0)
  })
})
